src/backend/joanie/tests/core/test_api_order.py

```diff
@@ -1912,3 +1969,383 @@ def test_api_order_create_no_limit(self):
+        """
+        Create an order on a free product
+        """
+        user = factories.UserFactory()
+        token = self.generate_token_from_user(user)
+        course = factories.CourseFactory()
+        product = factories.ProductFactory(courses=[course], price=0.00)
+        organization = product.course_relations.first().organizations.first()
+
+        data = {
+            "course": course.code,
+            "organization": str(organization.id),
+            "product": str(product.id),
+        }
+        response = self.client.post(
+            "/api/v1.0/orders/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(response.json()["state"], enums.ORDER_STATE_DRAFT)
+        order = models.Order.objects.get(id=response.json()["id"])
+        response = self.client.patch(
+            f"/api/v1.0/orders/{order.id}/submit/",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        order.refresh_from_db()
+        self.assertEqual(order.state, enums.ORDER_STATE_VALIDATED)
+
+    def test_api_order_create_no_billing_address_to_validation(self):
+        """
+        Create an order without billing address and continues the process
+        until validation
+        """
+        user = factories.UserFactory()
+        token = self.generate_token_from_user(user)
+        course = factories.CourseFactory()
+        product = factories.ProductFactory(courses=[course])
+        organization = product.course_relations.first().organizations.first()
+
+        data = {
+            "course": course.code,
+            "organization": str(organization.id),
+            "product": str(product.id),
+        }
+
+        response = self.client.post(
+            "/api/v1.0/orders/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(response.json()["state"], enums.ORDER_STATE_DRAFT)
+        order_id = response.json()["id"]
+        billing_address = BillingAddressDictFactory()
+        data["billing_address"] = billing_address
+        response = self.client.patch(
+            f"/api/v1.0/orders/{order_id}/submit/",
+            data=data,
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+        self.assertEqual(response.status_code, 201)
+        order = models.Order.objects.get(id=order_id)
+        self.assertEqual(order.state, enums.ORDER_STATE_SUBMITTED)
+
+        InvoiceFactory(order=order)
+        response = self.client.put(
+            f"/api/v1.0/orders/{order_id}/validate/",
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
```

Comment:
In real case, the order should be validated automatically once the payment notification has been received. So to stick to real case, I think you should call order.validate() instead to send a request.